libparsec/crates/platform_storage/src/web/certificates.rs

```diff
@@ -1,28 +1,121 @@
+];
+
+const REALM_CERTIFICATES: [&str; 4] = [
+    "realm_role_certificate",
+    "realm_name_certificate",
+    "realm_key_rotation_certificate",
+    "realm_archiving_certificate",
+];
+
+const SHAMIR_RECOVERY_CERTIFICATES: [&str; 2] = [
+    "shamir_recovery_share_certificate",
+    "shamir_recovery_brief_certificate",
+];
 
 #[derive(Debug)]
 pub(crate) struct PlatformCertificatesStorageForUpdateGuard<'a> {
-    _marker: PhantomData<&'a ()>,
+    transaction: IdbTransaction<'a>,
 }
 
 impl<'a> PlatformCertificatesStorageForUpdateGuard<'a> {
     pub async fn commit(self) -> anyhow::Result<()> {
-        todo!()
+        super::db::commit(self.transaction).await
     }
 
     pub async fn get_certificate_encrypted(
         &mut self,
         query: GetCertificateQuery,
         up_to: UpTo,
     ) -> Result<(DateTime, Vec<u8>), GetCertificateError> {
-        todo!()
+        // Handling `up_to` with a timestamp is a bit tricky:
+        // 1) we want the last result up to the given timestamp (included)
+        // 2) BUT ! if 1) fails we want the first result in order
+        //    to correctly return NonExisting vs ExistButTooRecent
+        //
+        // So to do that we act in three steps:
+        //  1) do the SQL query with the `up_to` filter
+        //  2) if the previous query returned nothing, do another SQL query without `up_to`
+        //  3) compare the datetime of the result with the on provided by `up_to`: if it's
+        //     too high it's a ExistButTooRecent error otherwise it's a valid result
+
+        // Step 1: do the regular SQL query
+        let certifs = Certificate::get_values(
+            &self.transaction,
+            CertificateFilter {
+                certificate_type: query.certificate_type.into(),
+                filter1: query.filter1.clone(),
+                filter2: query.filter2.clone(),
+            },
+        )
+        .await?;
+
+        let maybe_certif_timestamp = certifs
+            .get(0)
+            .map(|certif| (certif.certificate_timestamp, certif.certificate.clone()));
+
+        let certifs = if let UpTo::Timestamp(up_to) = up_to {
+            certifs
+                .into_iter()
+                .filter(|certif| certif.certificate_timestamp <= up_to.get_f64_with_us_precision())
+                .collect()
+        } else {
+            certifs
+        };
+
+        if let Some(certif) = certifs.into_iter().max_by(|x, y| {
+            x.certificate_timestamp
+                .partial_cmp(&y.certificate_timestamp)
+                .expect("Timestamp should not be undefined")
+        }) {
+            return Ok((
+                DateTime::from_f64_with_us_precision(certif.certificate_timestamp),
+                certif.certificate.to_vec(),
+            ));
+        }
+
+        let up_to = if let UpTo::Timestamp(up_to) = up_to {
+            up_to
+        } else {
+            return Err(GetCertificateError::NonExisting);
+        };
```

Comment:
```suggestion
        if !matches!(up_to, UpTo::Timestamp(_)) {
            return Err(GetCertificateError::NonExisting);
        };
```

I don't see `up_to` being used after, maybe `matches` is better here ?

Reply:
This is because I missed something:
```Rust
if certificate_timestamp > up_to {
    Err(GetCertificateError::ExistButTooRecent {
        certificate_timestamp,
    })
}
```

Reply:
Corrected